src/prototype_room_costmatrix.js

```diff
@@ -57,41 +57,47 @@ Room.prototype.setCostMatrixPath = function(costMatrix, path) {
     let pos = path[i];
     costMatrix.set(pos.x, pos.y, config.layout.pathAvoid);
   }
+  this.setCostMatrixAvoidSources(costMatrix);
```

Comment:
![source_avoid](https://user-images.githubusercontent.com/131156/28841481-ac8bfc34-76fa-11e7-8c72-bbb0f43346f4.png)

Maybe let's discuss at an example, feel free to provide better discussion material, my 'color'-skills are limited.
The red dots are the `setCostMatrixAvoidSources` positions. The green dot is the only position where the source avoid costmatrix entry gets overwritten, which you want to reset. I think positions on paths should be by default useable for paths. Positions close the sources are not the most interesting ones for pathing because the surrounding positions should be avoided, except you want to move to the source.

Regarding the special structure avoidance topic:
![structure_avoid](https://user-images.githubusercontent.com/131156/28841853-d555eaf2-76fb-11e7-810a-3502656102f0.png)
The yellow dots are the special structure avoidance. I guess the paths will look similar to the white ones. If I understand you correctly this is even more what you are trying to archive.

Why not have a gradient based on the range for special structures [SOURCE, MINERAL, SK_LAIR], which tries to avoid going close to these structures, but low enough to still use it if it makes sense.